### PR TITLE
Correlation of morphology and nomic data

### PR DESCRIPTION
**Pull request by MikeLippincott**
Here I correlate:
* Wells
* Treatments
* Selected treatments  
using bulk morphology and nomic nELISA data.

### CODE REVIEW

File: 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+save_path.mkdir(parents=True, exist_ok=True)
+well_corr_df.to_csv(f"{save_path}/treatments_corr.csv")
+
+
+# # Treatment correlation for selected treatments
+
+# In[14]:
+
+
+list_of_treatments = [
```
Comment (gwaybio) [inline 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py:203]: We set constants like this after package imports and function definitions. If you plan on using this list in the future, you may also consider adding it to a separate utils.py file and importing it

File: 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    "Thapsigargin_10.0_DMSO_0.025",
+]
+
+
+# In[15]:
+
+
+nomic_df = nomic_df.reset_index()
+# subset the data to only include the treatments of interest from list_of_treatments
+nomic_df = nomic_df[
```
Comment (gwaybio) [inline 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py:220]: Is there a reason to perform the subsetting now and not later? i.e. why not calculate the correlations for everything and then subset when you want to analyze them later (e.g., for plotting)

File: 1.Exploratory_Data_Analysis/scripts/2b.correlation_agg_morphology_features.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+# read in data
+feature_df = pd.read_parquet(feature_df_path)
+
+
+# ## Clean up df
+
+# In[4]:
+
+
+# remove uM in each row of the Metadata_inducer1_concentration column if it is present
```
Comment (gwaybio) [inline 1.Exploratory_Data_Analysis/scripts/2b.correlation_agg_morphology_features.py:46]: maybe i'm not reading closely enough, but it seems there is a lot of repeated code here. If you refactor to clean the data first, then you won't need to clean it every time.

File: 1.Exploratory_Data_Analysis/scripts/2b.correlation_agg_morphology_features.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+
+treatments_agg = feature_df
+treatments_agg = treatments_agg.reset_index()
+
+
+# In[11]:
+
+
+list_of_treatments = [
```
Comment (gwaybio) [inline 1.Exploratory_Data_Analysis/scripts/2b.correlation_agg_morphology_features.py:167]: it would be difficult to maintain this list in all three locations.

Also, consider adding a comment about why to focus on these treatments specifically.

File: 1.Exploratory_Data_Analysis/scripts/2c.correlation_nomic_and_agg_morphology.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+
+# merge nomic and morphology data by metadata_well and position_x
+merged_df = pd.merge(
+    feature_df,
+    nomic_df,
+    left_on=["Metadata_Well"],
+    right_on=["Metadata_position_x"],
+)
+merged_df = merged_df.drop(["Metadata_position_x"], axis=1)
```
Comment (gwaybio) [inline 1.Exploratory_Data_Analysis/scripts/2c.correlation_nomic_and_agg_morphology.py:142]:
```suggestion
merged_df = pd.merge(
    feature_df,
    nomic_df,
    left_on=["Metadata_Well"],
    right_on=["Metadata_position_x"],
).drop(["Metadata_position_x"], axis=1)
```

Welcome to pandas chaining!

File: 1.Exploratory_Data_Analysis/scripts/2c.correlation_nomic_and_agg_morphology.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+
+
+# ## Correlation of Aggregated Features per Well + Nomic Data
+
+# In[8]:
+
+
+# drop metadata columns if the column contains 'Metadata'
+cell_df = merged_df
```
Comment (gwaybio) [inline 1.Exploratory_Data_Analysis/scripts/2c.correlation_nomic_and_agg_morphology.py:151]: not sure if this is necessary? why not just use `merged_df` directly?

File: 1.Exploratory_Data_Analysis/scripts/2c.correlation_nomic_and_sc_morphology.py [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+
+treatment = "DMSO_0.100_DMSO_0.025"
+# treatment = 'DMSO_0.100_Z-VAD-FMK_100.0'
+print(treatment)
+# get the row
+tmpdf = merged_data_agg.loc[
+    merged_data_agg["oneb_Metadata_Treatment_Dose_Inhibitor_Dose"] == treatment
+]
+tmpdf
+# get the correlation between the sc data and nomic data for well B13
```
Comment (gwaybio) [inline 1.Exploratory_Data_Analysis/scripts/2c.correlation_nomic_and_sc_morphology.py:164]: Is it worth deleting this vestigial code?

File: 1.Exploratory_Data_Analysis/scripts/2c.correlation_nomic_and_sc_morphology.py [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+    tmpdf = tmpdf.rename(columns={"index": "feature"})
+    # save the correlation matrix to a csv
+
+    # # melt the correlation matrix
+    corr_matrix_melted = tmpdf.reset_index().melt(
+        id_vars="feature", value_vars=tmpdf.columns[1:]
+    )
+    corr_matrix_melted = corr_matrix_melted.rename(
+        columns={"index": "feature", "variable": "feature2", "value": "correlation"}
+    )
```
Comment (gwaybio) [inline 1.Exploratory_Data_Analysis/scripts/2c.correlation_nomic_and_sc_morphology.py:236]: All of this can be chained, which makes for a much more pleasant reading experience :)

Can you also think of a more readable name than `tmpdf`?

File: 1.Exploratory_Data_Analysis/scripts/2c.correlation_nomic_and_sc_morphology.py [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+
+    # # find the treatment dose for well B13 in sc_data_metadata
+
+    corr_matrix_melted["treatment"] = treatment
+    corr_matrix_melted.to_csv(f"{cell_type}_{treatment}_correlation_matrix.csv")
+    # filter out the correlation values that are less than abs 0.5
+    corr_matrix_melted = corr_matrix_melted.loc[
+        abs(corr_matrix_melted["correlation"]) > 0.5
```
Comment (gwaybio) [inline 1.Exploratory_Data_Analysis/scripts/2c.correlation_nomic_and_sc_morphology.py:244]: hmm, why are we throwing away data?

File: 1.Exploratory_Data_Analysis/scripts/3.correlation_visualization.r [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    setwd(figure_dir)
+}
+print(figure_dir)
+
+# plot corr plot for all wells
+# set the file name
+file = paste0(group,".png")
+print(file)
+png(height=1800, width=1800, file=(file.path(figure_dir,file)), type = "cairo")
+corrplot(df,
```
Comment (gwaybio) [inline 1.Exploratory_Data_Analysis/scripts/3.correlation_visualization.r:59]: These corrplots are cool but very hard to read (especially the ones with a lot of data)

Is there an alternative way to visualize these data? I know it is a lot to digest.

Essentially, we need to visualize five things:

1. Treatment A
2. Treatment B
3. Dose
4. Correlation
5. Relationship between dose and correlation

There are formal strategies for deciding the figure type, but I cannot find the resource presently.

We can chat strategy in this discussion or next week

**Review by gwaybio — APPROVED** [review]
I made several in-line comments, mostly about style.

They are not required changes as they won't likely impact results, but consider implementing them as they would demonstrate increased code efficiency, readability, and growth. Well done!

File: 1.Exploratory_Data_Analysis/correlation_run.sh [hunk 11]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 1.Exploratory_Data_Analysis/correlation_run.sh:3]: Added!

File: 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+)
+
+# read in data
+nomic_df_raw = pd.read_csv(nomic_df_path)
+
+
+# In[4]:
+
+
+# remove column if colname has pgml in it
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py:46]: Addressed

File: 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+# # Normalization of Values
+
+# In[7]:
+
+
+# min-max normalization of nomic data from scipy
+from sklearn.preprocessing import MinMaxScaler
+
+scaler = MinMaxScaler()
+nomic_df = pd.DataFrame(scaler.fit_transform(nomic_df), columns=nomic_df.columns)
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py:140]: @gwaybio Yes, I decided to normalize for each cell line independently.

File: 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+
+# summary statistics of df to check min-max normalization
+nomic_df.describe()
+
+
+# In[9]:
+
+
+# add position_x back to df
+nomic_df.loc[:, "Metadata_position_x"] = nomic_df_raw["position_x"]
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py:154]: addressed

File: 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py (hunk 1 above)
Comment (MikeLippincott) [reply 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py:203]: addressed via adding a `params.py` file for import

File: 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py (hunk 2 above)
Comment (MikeLippincott) [reply 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py:220]: Good point, I have not thought about this actually. Makes more sense when using tidy long format to avoid duplication of groups. I will change this in this PR.

File: 1.Exploratory_Data_Analysis/scripts/2b.correlation_agg_morphology_features.py (hunk 3 above)
Comment (MikeLippincott) [reply 1.Exploratory_Data_Analysis/scripts/2b.correlation_agg_morphology_features.py:46]: added '0.nomic_cleanup.ipynb` to adress this comment

File: 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+# if column does not contain [NSU] then prefix with Metadata_
+for col in nomic_df.columns:
+    if not any(x in col for x in ["NSU"]):
+        nomic_df = nomic_df.rename(columns={col: "Metadata_" + col})
+
+
+# In[5]:
+
+
+## Clean up df
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 1.Exploratory_Data_Analysis/scripts/2a.correlation_nomic_features.py:59]: addressed and added!

File: 1.Exploratory_Data_Analysis/scripts/3.correlation_visualization.r (hunk 10 above)
Comment (MikeLippincott) [reply 1.Exploratory_Data_Analysis/scripts/3.correlation_visualization.r:59]: We will definitely need to talk more about data visualization.
